Plutarch/FFI.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.FFI (
+  foreignExport,
+  foreignImport,
+) where
+
+import Data.Kind (Constraint, Type)
+import Data.Text (Text)
+import GHC.TypeLits (TypeError)
+import qualified GHC.TypeLits as TypeLits
+import qualified Generics.SOP as SOP
+import Plutarch.Bool (PBool)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  ClosedTerm,
+  PDelayed,
+  PType,
+  RawTerm (RCompiled),
+  Term (Term),
+  TermResult (TermResult),
+  asClosedRawTerm,
+  compile',
+  (:-->),
+ )
+import Plutarch.Internal.PlutusType (PlutusType (PInner))
+import Plutarch.String (PString)
+import Plutus.V1.Ledger.Scripts (Script (unScript), fromCompiledCode)
+import PlutusTx.Code (CompiledCode, CompiledCodeIn (DeserializedCode))
+import PlutusTx.Prelude (BuiltinString)
+import UntypedPlutusCore (fakeNameDeBruijn)
+import qualified UntypedPlutusCore as UPLC
+
+data ForallPhantom :: Type
+data PhorallPhantom :: PType
+
+foreignExport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => ClosedTerm p -> CompiledCode t
+foreignExport t = DeserializedCode program Nothing mempty
+  where
+    program =
+      UPLC.Program () (UPLC.Version () 1 0 0) $
+        UPLC.termMapNames fakeNameDeBruijn $
+          compile' $
+            asClosedRawTerm t
+
+foreignImport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => CompiledCode t -> ClosedTerm p
+foreignImport c = Term $ const $ TermResult (RCompiled $ UPLC.toTerm $ unScript $ fromCompiledCode c) []
+
+type family a ~~ b :: Constraint where
+  ForallPhantom ~~ _ = ()
+  _ ~~ ForallPhantom = ()
+  a ~~ b = a ~ b
+
+type family PlutarchInner (p :: PType) (any :: PType) :: Type where
+  PlutarchInner PBool _ = Bool
+  PlutarchInner PInteger _ = Integer
+  PlutarchInner PString _ = Text
+  PlutarchInner PhorallPhantom _ = ForallPhantom
+  PlutarchInner (a :--> b) x = PlutarchInner a b -> PlutarchInner b x  -- hack to support Scott encodings
+  PlutarchInner (PDelayed a) x = PlutarchInner a x
+  PlutarchInner p x = PlutarchInner (PInner p x) x
+
+type family PlutusTxInner (t :: Type) (any :: Type) :: Type where
+  PlutusTxInner Bool _ = Bool
+  PlutusTxInner Integer _ = Integer
+  PlutusTxInner BuiltinString _ = Text
+  PlutusTxInner ForallPhantom _ = ForallPhantom
+  PlutusTxInner (a -> b) x = PlutusTxInner a b -> PlutusTxInner b x   -- hack to support Scott encodings
+  PlutusTxInner a x = PlutusTxInner (ScottFn (ScottList (SOP.Code a) x) x) x
```

Comment:
Doesn't `ScottList (ToPType2 (SOP.Code a))` allow us to reuse the existing type families from `PlutusType.hs`?